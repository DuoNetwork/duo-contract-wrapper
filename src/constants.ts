export const ETH_MAINNET_ID = 1;
export const ETH_KOVAN_ID = 42;

export const DUMMY_ADDR = '0x0';

export const CTD_INCEPTION = 'Inception';
export const CTD_TRADING = 'Trading';
export const CTD_PRERESET = 'PreReset';
export const CTD_RESET = 'Reset';
export const CTD_LOADING = 'Loading';

export const BTV_UP_RESET = 'UpwardReset';
export const BTV_DOWN_RESET = 'DownwardReset';
export const BTV_PERIOD_RESET = 'PeriodicReset';

export const ESP_NOT_STARTED = 'NotStarted';
export const ESP_MODERATOR = 'Moderator';
export const ESP_CONTRACT = 'Contract';

// wallet and contract
export const INCEPTION_BLK_KOVAN = 9355871;
export const INCEPTION_BLK_MAIN = 0;

// managed
export const EVENT_UPDATE_ROLE_MANAGER = 'UpdateRoleManager';
export const EVENT_UPDATE_OPERATOR = 'UpdateOperator';

// esplanade
export const EVENT_ADD_ADDRESS = 'AddAddress';
export const EVENT_REMOVE_ADDRESS = 'RemoveAddress';
export const EVENT_PROVIDE_ADDRESS = 'ProvideAddress';
export const EVENT_ADD_CUSTODIAN = 'AddCustodian';
export const EVENT_ADD_OTHER_CONTRACT = 'AddOtherContract';
export const EVENT_START_CONTRACT_VOTING = 'StartContractVoting';
export const EVENT_TERMINATE_CONTRACT_VOTING = 'TerminateContractVoting';
export const EVENT_START_MODERATOR_VOTING = 'StartModeratorVoting';
export const EVENT_TERMINATEBY_TIMEOUT = 'TerminateByTimeOut';
export const EVENT_VOTE = 'Vote';
export const EVENT_COMPLETE_VOTING = 'CompleteVoting';
export const EVENT_REPLACE_MODERATOR = 'ReplaceModerator';

// custodian
export const EVENT_START_TRADING = 'StartTrading';
export const EVENT_START_PRE_RESET = 'StartPreReset';
export const EVENT_START_RESET = 'StartReset';
export const EVENT_MATURED = 'Matured'
export const EVENT_ACCEPT_PRICE = 'AcceptPrice';
export const EVENT_CREATE = 'Create';
export const EVENT_REDEEM = 'Redeem';
export const EVENT_TOTAL_SUPPLY = 'TotalSupply';
export const EVENT_TRANSFER = 'Transfer';
export const EVENT_APPROVAL = 'Approval';
export const EVENT_COLLECT_FEE = 'CollectFee';
export const EVENT_UPDATE_ORACLE = 'UpdateOracle';
export const EVENT_UPDATE_FEE_COLLECTOR = 'UpdateFeeCollector';

// beethovan
export const EVENT_SET_VALUE = 'SetValue';

// magi
export const EVENT_COMMIT_PRICE = 'CommitPrice';
export const EVENT_UPDATE_PRICE_FEED = 'UpdatePriceFeed';

export const STATE_INCEPTION = '0';
export const STATE_TRADING = '1';
export const STATE_PRERESET = '2';
export const STATE_RESET = '3';
export const RESET_STATE_UP = '0';
export const RESET_STATE_DOWN = '1';
export const RESET_STATE_PERIOD = '2';
export const VOTING_NOT_STARTED = '0';
export const VOTING_MODERATOR = '1';
export const VOTING_CONTRACT = '2';
export const FN_START_CONTRACT = 'startContract';
export const FN_COMMIT_PRICE = 'commitPrice';
export const SRC_MYETHER = 'myether';
export const SRC_INFURA = 'infura';
export const PROVIDER_LOCAL_HTTP = 'http://localhost:8545';
// export const PROVIDER_LOCAL_WS = 'ws://localhost:8546';
export const PROVIDER_INFURA_MAIN_WS = 'wss://mainnet.infura.io/ws';
export const PROVIDER_INFURA_KOVAN_WS = 'wss://kovan.infura.io/ws';
export const PROVIDER_MYETHER_MAIN = 'https://api.myetherapi.com/eth';
export const PROVIDER_MYETHER_ROP = 'https://api.myetherapi.com/rop';
export const PROVIDER_INFURA_MAIN = 'https://mainnet.infura.io';
export const PROVIDER_INFURA_KOVAN = 'https://kovan.infura.io';
export const TRANSFER_TOKEN_INTERVAL = 30; // in seconds
export const TRANSFER_TOKEN_GAS = 120000;
export const TRANSFER_TOKEN_GAS_TH = 0.01; // in ether
export const SET_VALUE_GAS = 50000;
export const SET_VALUE_GAS_PRICE = 3; // in Gwei
export const COLLECT_FEE_GAS = 40000;
export const COLLECT_FEE_GAS_PRICE = 4; // in Gwei
export const ADD_ADDR_GAS = 40000;
export const ADD_ADDR_GAS_PRICE = 5; // in Gwei
export const REMOVE_ADDR_GAS = 40000;
export const REMOVE_ADDR_GAS_PRICE = 5; // in Gwei
export const UPDATE_ADDR_GAS = 40000;
export const UPDATE_ADDR_GAS_PRICE = 5; // in Gwei
export const TRANSFER_INTERVAL = 10; // in seconds
export const TRANSFER_GAS_TH = 0.01;
export const REDEEM_INTERVAL = 10; // in seconds
export const REDEEM_GAS = 80000;
export const REDEEM_GAS_TH = 0.005;
export const CREATE_INTERVAL = 10; // in seconds
export const CREATE_GAS = 160000;
export const CREATE_GAS_TH = 0.01;
export const DEFAULT_GAS_PRICE = 5;
export const PRE_RESET_GAS_LIMIT = 120000;
export const RESET_GAS_LIMIT = 4000000;
export const EVENT_FETCH_BLOCK_INTERVAL = 100;
export const EVENT_FETCH_TIME_INVERVAL = 600000;

export const LOG_INFO = 'INFO';
export const LOG_DEBUG = 'DEBUG';
export const LOG_ERROR = 'ERROR';
export const LOG_RANKING: { [level: string]: number } = {
	[LOG_ERROR]: 0,
	[LOG_INFO]: 1,
	[LOG_DEBUG]: 2
};

export const BTV_STATE = {
	LAST_OPERATION_TIME: 0,
	OPERATION_COOLDOWN: 1,
	STATE: 2,
	MIN_BALANCE: 3,
	TOTAL_SUPPLYA: 4,
	TOTAL_SUPPLYB: 5,
	ETH_COLLATERAL_INWEI: 6,
	NAVA_INWEI: 7,
	NAVB_INWEI: 8,
	LAST_PRICE_INWEI: 9,
	LAST_PRICETIME_INSECOND: 10,
	RESET_PRICE_INWEI: 11,
	RESET_PRICETIME_INSECOND: 12,
	CREATE_COMMINBP: 13,
	REDEEM_COMMINBP: 14,
	PERIOD: 15,
	MATURITY: 16,
	PRERESET_WAITING_BLOCKS: 17,
	PRICE_FETCH_COOLDOWN: 18,
	NEXT_RESET_ADDR_INDEX: 19,
	TOTAL_USERS: 20,
	FEE_BALANCE_INWEI: 21,
	RESET_STATE: 22,
	ALPHA_INBP: 23,
	BETA_INWEI: 24,
	PERIOD_COUPON_INWEI: 25,
	LIMIT_PERIODIC_INWEI: 26,
	LIMIT_UPPER_INWEI: 27,
	LIMIT_LOWER_INWEI: 28,
	ITERATION_GAS_THRESHOLD: 29
};
