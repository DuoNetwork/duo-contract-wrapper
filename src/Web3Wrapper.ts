import * as CST from './constants';
import { kovan, mainnet } from './contractAddresses';
import erc20Abi from './static/ERC20.json';
import { IContractAddresses, IEvent, ITransactionOption, Wallet } from './types';

const Web3 = require('web3');
const HDWalletProvider = require('./external/HDWalletProvider');
const BigNumber = require('bignumber.js');
const ProviderEngine = require('web3-provider-engine');
const FetchSubprovider = require('web3-provider-engine/subproviders/fetch');
const createLedgerSubprovider = require('@ledgerhq/web3-subprovider').default;
const TransportU2F = require('@ledgerhq/hw-transport-u2f').default;

export class Web3Wrapper {
	public web3: any;
	public wallet: Wallet = Wallet.None;
	public accountIndex: number = 0;
	private live: boolean;
	private providerUrl: string;
	public readonly contractAddresses: IContractAddresses;
	public readonly inceptionBlockNumber: number;
	public handleSwitchToMetaMask: Array<() => any>;
	public handleSwitchToLedger: Array<() => any>;

	constructor(window: any, providerUrl: string, privateKey: string, live: boolean) {
		this.live = live;
		this.contractAddresses = this.live ? mainnet : kovan;
		this.providerUrl = providerUrl;
		if (window && (window.ethereum || window.web3)) {
			this.web3 = new Web3(window.ethereum || window.web3.currentProvider);
			this.wallet = Wallet.MetaMask;
		} else if (!window && privateKey) {
			const hdWallet = new HDWalletProvider(privateKey, providerUrl);
			this.web3 = new Web3(hdWallet);
			this.wallet = Wallet.Local;
		} else {
			this.web3 = new Web3(providerUrl);
			this.wallet = Wallet.None;
		}
		this.handleSwitchToMetaMask = [];
		this.handleSwitchToLedger = [];
		this.inceptionBlockNumber = live ? CST.INCEPTION_BLK_MAIN : CST.INCEPTION_BLK_KOVAN;
	}

	public onSwitchToMetaMask(handleSwitchToMetaMask: () => any) {
		this.handleSwitchToMetaMask.push(handleSwitchToMetaMask);
	}

	public onSwitchToLedger(handleSwitchToLedger: () => any) {
		this.handleSwitchToLedger.push(handleSwitchToLedger);
	}

	public switchToMetaMask(window: any) {
		if (window && (window.ethereum || window.web3)) {
			this.web3 = new Web3(window.ethereum || window.web3.currentProvider);
			this.wallet = Wallet.MetaMask;
		} else {
			this.web3 = new Web3(this.providerUrl);
			this.wallet = Wallet.None;
		}
		this.accountIndex = 0;
		this.handleSwitchToMetaMask.forEach(h => {
			try {
				h();
			} catch (error) {
				console.log(error);
			}
		});
	}

	public async switchToLedger() {
		const engine = new ProviderEngine();
		const getTransport = () => TransportU2F.create();
		const networkId = this.live ? CST.ETH_MAINNET_ID : CST.ETH_KOVAN_ID;
		const rpcUrl = this.providerUrl;
		const ledger = createLedgerSubprovider(getTransport, {
			networkId,
			accountsLength: 20
		});
		engine.addProvider(ledger);
		engine.addProvider(new FetchSubprovider({ rpcUrl }));
		engine.start();
		const newWeb3 = new Web3(engine);
		const accounts = await newWeb3.eth.getAccounts();
		this.web3 = newWeb3;
		this.wallet = Wallet.Ledger;

		this.handleSwitchToLedger.forEach(h => {
			try {
				h();
			} catch (error) {
				console.log(error);
			}
		});
		return accounts;
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		if (this.wallet !== Wallet.MetaMask) return;

		const store = (this.web3.currentProvider as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				if (
					this.wallet === Wallet.MetaMask &&
					store.getState().selectedAddress &&
					store.getState().networkVersion
				)
					onUpdate(
						store.getState().selectedAddress,
						Number(store.getState().networkVersion)
					);
			});
	}

	public isReadOnly() {
		return this.wallet === Wallet.None;
	}

	public readOnlyReject() {
		return Promise.reject('Read Only Mode');
	}

	public isLocal() {
		return this.wallet === Wallet.Local;
	}

	public isLive() {
		return this.live;
	}

	public wrongEnvReject() {
		return Promise.reject('Wrong Env');
	}

	public getGasPrice(): Promise<string> {
		return this.web3.eth.getGasPrice();
	}

	public async sendEther(
		from: string,
		to: string,
		value: number,
		option: ITransactionOption = {}
	): Promise<any> {
		const gasPrice = option.gasPrice || (await this.getGasPrice());
		const gasLimit = option.gasLimit || CST.DEFAULT_TX_GAS_LIMIT;
		const nonce = option.nonce || (await this.getTransactionCount(from));

		return this.web3.eth.sendTransaction({
			from: from,
			to: to,
			value: Web3Wrapper.toWei(value),
			gasPrice: gasPrice,
			gas: gasLimit,
			nonce: nonce
		});
	}

	public async erc20Transfer(
		contractAddress: string,
		from: string,
		to: string,
		value: number,
		option: ITransactionOption = {}
	): Promise<any> {
		if (this.isReadOnly()) return this.readOnlyReject();
		const gasPrice = option.gasPrice || (await this.getGasPrice());
		const gasLimit = option.gasLimit || CST.DEFAULT_TX_GAS_LIMIT;
		const nonce = option.nonce || (await this.getTransactionCount(from));
		const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);

		return erc20Contract.methods.transfer(to, Web3Wrapper.toWei(value)).send({
			from: from,
			gasPrice: gasPrice,
			gas: gasLimit,
			nonce: nonce
		});
	}

	public async erc20Approve(
		contractAddress: string,
		from: string,
		spender: string,
		value: number,
		unlimited: boolean = false
	): Promise<any> {
		if (this.isReadOnly()) return this.readOnlyReject();
		return new Promise<string>(resolve => {
			const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);
			return erc20Contract.methods
				.approve(
					spender,
					unlimited
						? Web3Wrapper.toHex(new BigNumber(2).pow(256).minus(1))
						: Web3Wrapper.toWei(value)
				)
				.send({
					from: from
				})
				.on('transactionHash', (txHash: string) => resolve(txHash));
		});
	}

	public getCurrentBlockNumber(): Promise<number> {
		return this.web3.eth.getBlockNumber();
	}

	public getBlock(
		blkNumber: number | string
	): Promise<{
		number: number;
		hash: string;
		parentHash: string;
		nonce: string;
		sha3Uncles: string;
		logsBloom: string;
		transactionsRoot: string;
		stateRoot: string;
		receiptsRoot: string;
		miner: string;
		difficulty: string;
		totalDifficulty: string;
		size: number;
		extraData: string;
		gasLimit: string;
		gasUsed: string;
		timestamp: number;
		transactions: string[];
		uncles: string[];
	}> {
		return this.web3.eth.getBlock(blkNumber);
	}

	public async getBlockTimestamp(blkNumber: number = 0) {
		const blk = await this.getBlock(blkNumber || 'latest');
		if (!blk)
			throw new Error('Error: no block found for ' + blkNumber);
		return blk.timestamp * 1000;
	}

	public async getCurrentAddress(): Promise<string> {
		const accounts = await this.web3.eth.getAccounts();
		return accounts[this.accountIndex] || CST.DUMMY_ADDR;
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3.eth.net.getId();
	}

	public async getEthBalance(address: string): Promise<number> {
		return Web3Wrapper.fromWei(await this.web3.eth.getBalance(address));
	}

	public async getErc20Balance(contractAddress: string, address: string) {
		const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);

		return Web3Wrapper.fromWei(await erc20Contract.methods.balanceOf(address).call());
	}

	public async getErc20Allowance(contractAddress: string, address: string, spender: string) {
		const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);

		return Web3Wrapper.fromWei(await erc20Contract.methods.allowance(address, spender).call());
	}

	public static fromWei(value: string | number) {
		return Number(Web3.utils.fromWei(value, 'ether'));
	}

	public static toWei(value: string | number): any {
		return Web3.utils.toWei(value + '', 'ether');
	}

	public static toHex(value: any): string {
		return Web3.utils.toHex(value);
	}

	public static checkAddress(addr: string): boolean {
		if (!addr.startsWith('0x') || addr.length !== 42) return false;
		return Web3.utils.checkAddressChecksum(Web3.utils.toChecksumAddress(addr));
	}

	public getTransactionReceipt(
		txHash: string
	): Promise<null | {
		status: boolean;
		blockHash: string;
		blockNumber: number;
		transactionHash: string;
		transactionIndex: number;
		from: string;
		to: string;
		contractAddress: string | null;
		cumulativeGasUsed: number;
		gasUsed: number;
		logs: Array<{
			data: string;
			topics: string[];
			logIndex: number;
			transactionIndex: number;
			transactionHash: string;
			blockHash: string;
			blockNumber: number;
			address: string;
		}>;
	}> {
		return this.web3.eth.getTransactionReceipt(txHash);
	}

	public static parseEvent(
		eventLog: {
			event: string;
			address: string;
			returnValues: any;
			logIndex: number;
			transactionIndex: number;
			transactionHash: string;
			blockHash: string;
			blockNumber: number;
			raw?: { data: string; topics: string[] };
		},
		timestamp: number
	): IEvent {
		const returnValue = eventLog.returnValues;
		const output: IEvent = {
			contractAddress: eventLog.address,
			type: eventLog.event,
			id: (eventLog as any)['id'],
			blockHash: eventLog.blockHash,
			blockNumber: eventLog.blockNumber,
			transactionHash: eventLog.transactionHash,
			logStatus: (eventLog as any)['type'],
			parameters: {},
			timestamp: timestamp
		};
		for (const key in returnValue) output.parameters[key] = returnValue[key];

		return output;
	}

	public static pullEvents(
		contract: any,
		start: number,
		end: number,
		event: string
	): Promise<
		Array<{
			event: string;
			address: string;
			returnValues: any;
			logIndex: number;
			transactionIndex: number;
			transactionHash: string;
			blockHash: string;
			blockNumber: number;
			raw?: { data: string; topics: string[] };
		}>
	> {
		return contract.getPastEvents(event, {
			fromBlock: start,
			toBlock: end
		});
	}

	public createContract(abi: any[], address: string): any {
		return new this.web3.eth.Contract(abi, address);
	}

	public getTransactionCount(address: string): Promise<number> {
		return this.web3.eth.getTransactionCount(address);
	}

	public async getTransactionOption(
		account: string,
		defaultGasLimit: number,
		option: ITransactionOption = {}
	) {
		const gasPrice = option.gasPrice || (await this.getGasPrice());
		const gasLimit = option.gasLimit || defaultGasLimit;
		const nonce = option.nonce || (await this.getTransactionCount(account));
		return {
			from: account,
			gasPrice: gasPrice,
			gas: gasLimit,
			nonce: nonce
		};
	}
}

export default Web3Wrapper;
