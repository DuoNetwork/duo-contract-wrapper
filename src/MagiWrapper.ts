import BaseContractWrapper from './BaseContractWrapper';
import * as CST from './constants';
import magiAbi from './static/Magi.json';
import { IContractPrice, IEthTxOption } from './types';
import Web3Wrapper from './Web3Wrapper';

export default class MagiWrapper extends BaseContractWrapper {
	public readonly events = [
		CST.EVENT_UPDATE_ROLE_MANAGER,
		CST.EVENT_UPDATE_OPERATOR,
		CST.EVENT_COMMIT_PRICE,
		CST.EVENT_ACCEPT_PRICE,
		CST.EVENT_SET_VALUE,
		CST.EVENT_UPDATE_PRICE_FEED
	];
	constructor(web3Wrapper: Web3Wrapper, address: string) {
		super(web3Wrapper, magiAbi.abi, address);
	}

	public async isStarted(): Promise<boolean> {
		return await this.contract.methods.started().call();
	}

	public async startMagi(
		account: string,
		price: number,
		timeInSecond: number,
		option: IEthTxOption = {}
	) {
		if (!this.web3Wrapper.isLocal()) return this.web3Wrapper.wrongEnvReject();
		const gasPrice = option.gasPrice || (await this.web3Wrapper.getGasPrice());
		const gasLimit = option.gasLimit || CST.START_MAGI_GAS;
		return this.contract.methods.startOracle(this.web3Wrapper.toWei(price), timeInSecond).send({
			from: account || (await this.web3Wrapper.getCurrentAddress()),
			gasPrice: gasPrice,
			gas: gasLimit
		});
	}

	public async commitPrice(
		account: string,
		price: number,
		timeInSecond: number,
		option: IEthTxOption = {}
	) {
		if (!this.web3Wrapper.isLocal()) return this.web3Wrapper.wrongEnvReject();
		const gasPrice = option.gasPrice || (await this.web3Wrapper.getGasPrice());
		const gasLimit = option.gasLimit || CST.COMMIT_PRICE_GAS;
		return this.contract.methods.commitPrice(this.web3Wrapper.toWei(price), timeInSecond).send({
			from: account || (await this.web3Wrapper.getCurrentAddress()),
			gasPrice: gasPrice,
			gas: gasLimit
		});
	}

	public async getLastPrice(): Promise<IContractPrice> {
		const lastPrice = await this.contract.methods.getLastPrice().call();
		return {
			price: this.web3Wrapper.fromWei(lastPrice[0].valueOf()),
			timestamp: Number(lastPrice[1].valueOf()) * 1000
		};
	}
}
