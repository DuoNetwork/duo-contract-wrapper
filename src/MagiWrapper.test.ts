// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import * as CST from './constants';
import MagiWrapper from './MagiWrapper';
import Web3Wrapper from './Web3Wrapper';

const web3Wrapper = new Web3Wrapper(null, CST.PROVIDER_INFURA_KOVAN, '', false);
const magiWrapper = new MagiWrapper(web3Wrapper, web3Wrapper.contractAddresses.Oracles[0].address);

test('startMagi, wrong env', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => false);
	try {
		await magiWrapper.startMagi('address', 100, 123456789);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
});

test('startMagi, without option', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.contract.methods.startOracle = jest.fn(() => ({
		send: jest.fn()
	}));
	await magiWrapper.startMagi('address', 100, 123456789);
	expect((magiWrapper.contract.methods.startOracle as jest.Mock).mock.calls).toMatchSnapshot();

	// expect((magiWrapper.commitInternal as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startMagi, with option', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.contract.methods.startOracle = jest.fn(() => ({
		send: jest.fn()
	}));
	await magiWrapper.startMagi('address', 100, 123456789, {
		gasPrice: 2000000000,
		gasLimit: 200000
	});
	expect((magiWrapper.contract.methods.startOracle as jest.Mock).mock.calls).toMatchSnapshot();
});

test('startMagi, without account', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.web3Wrapper.getCurrentAddress = jest.fn(() =>
		Promise.resolve('0x0017d61f0B0a28E2F0eBB3B6E269738a6252CFeD')
	);
	magiWrapper.contract.methods.startOracle = jest.fn(() => ({
		send: jest.fn()
	}));
	await magiWrapper.startMagi('', 100, 123456789, {
		gasPrice: 2000000000,
		gasLimit: 200000
	});
	expect((magiWrapper.contract.methods.startOracle as jest.Mock).mock.calls).toMatchSnapshot();
});

test('commitPrice, wrong env', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => false);
	try {
		await magiWrapper.commitPrice('address', 100, 123456789);
	} catch (err) {
		expect(err).toMatchSnapshot();
	}
});

test('commitPrice, without option', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.contract.methods.commitPrice = jest.fn(() => ({
		send: jest.fn()
	}));
	await magiWrapper.commitPrice('address', 100, 123456789);
	expect((magiWrapper.contract.methods.commitPrice as jest.Mock).mock.calls).toMatchSnapshot();

	// expect((magiWrapper.commitInternal as jest.Mock).mock.calls).toMatchSnapshot();
});

test('commitPrice, with option', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.contract.methods.commitPrice = jest.fn(() => ({
		send: jest.fn()
	}));
	await magiWrapper.commitPrice('address', 100, 123456789, {
		gasPrice: 2000000000,
		gasLimit: 200000
	});
	expect((magiWrapper.contract.methods.commitPrice as jest.Mock).mock.calls).toMatchSnapshot();
});

test('commitPrice, without account', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.web3Wrapper.getCurrentAddress = jest.fn(() =>
		Promise.resolve('0x0017d61f0B0a28E2F0eBB3B6E269738a6252CFeD')
	);
	magiWrapper.contract.methods.commitPrice = jest.fn(() => ({
		send: jest.fn()
	}));
	await magiWrapper.commitPrice('', 100, 123456789, {
		gasPrice: 2000000000,
		gasLimit: 200000
	});
	expect((magiWrapper.contract.methods.commitPrice as jest.Mock).mock.calls).toMatchSnapshot();
});

test('getLastPrice', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.web3Wrapper.getCurrentAddress = jest.fn(() =>
		Promise.resolve('0x0017d61f0B0a28E2F0eBB3B6E269738a6252CFeD')
	);
	magiWrapper.contract.methods.getLastPrice = jest.fn(() => ({
		call: jest.fn(() => Promise.resolve(
			['100', '123456789']
		))
	}));
	const res = await magiWrapper.getLastPrice();
	expect((magiWrapper.contract.methods.getLastPrice as jest.Mock).mock.calls).toMatchSnapshot();
	expect(res).toMatchSnapshot();
});

test('isStarted', async () => {
	magiWrapper.web3Wrapper.isLocal = jest.fn(() => true);
	magiWrapper.web3Wrapper.getGasPrice = jest.fn(() => Promise.resolve(1000000000));
	magiWrapper.web3Wrapper.getCurrentAddress = jest.fn(() =>
		Promise.resolve('0x0017d61f0B0a28E2F0eBB3B6E269738a6252CFeD')
	);
	magiWrapper.contract.methods.isStarted = jest.fn(() => ({
		call: jest.fn(() => Promise.resolve(
			false
		))
	}));
	const res = await magiWrapper.isStarted();
	expect((magiWrapper.contract.methods.isStarted as jest.Mock).mock.calls).toMatchSnapshot();
	expect(res).toMatchSnapshot();
});
