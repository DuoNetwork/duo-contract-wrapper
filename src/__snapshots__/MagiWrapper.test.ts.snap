// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commitPrice, with option 1`] = `
Array [
  Array [
    "100000000000000000000",
    123456789,
  ],
]
`;

exports[`commitPrice, without account 1`] = `
Array [
  Array [
    "100000000000000000000",
    123456789,
  ],
]
`;

exports[`commitPrice, without option 1`] = `
Array [
  Array [
    "100000000000000000000",
    123456789,
  ],
]
`;

exports[`commitPrice, wrong env 1`] = `"Wrong Env"`;

exports[`getLastPrice 1`] = `
Array [
  Array [],
]
`;

exports[`getLastPrice 2`] = `
Object {
  "price": 1e-16,
  "timestamp": 123456789000,
}
`;

exports[`isStarted 1`] = `Array []`;

exports[`isStarted 2`] = `true`;

exports[`startMagi, with option 1`] = `
Array [
  Array [
    "100000000000000000000",
    123456789,
  ],
]
`;

exports[`startMagi, without account 1`] = `
Array [
  Array [
    "100000000000000000000",
    123456789,
  ],
]
`;

exports[`startMagi, without option 1`] = `
Array [
  Array [
    "100000000000000000000",
    123456789,
  ],
]
`;

exports[`startMagi, wrong env 1`] = `"Wrong Env"`;
